### PR TITLE
replace firestore.WriteBatch with DocUpdate

### PR DESCRIPTION
**Pull request by danrubel**
As another step towards [removing firebase from the model](https://github.com/larksystems/Project-2019-WorldBank-PLR/issues/31), this PR replaces `firestore.WriteBatch` with `DocUpdate`.

### CODE REVIEW

File: webapp/lib/model.g.dart [hunk 1]
```diff
@@ -417,12 +417,43 @@ void listenForUpdates<T>(
...
   final Map<String, dynamic> data;
 
   DocSnapshot(this.id, this.data);
 }
 
+/// A batch update, used to perform multiple writes as a single atomic unit.
+/// None of the writes are committed (or visible locally) until
+/// [DocUpdate.commit()] is called.
+abstract class DocUpdate {
+  /// Commits all of the writes in this write batch as a single atomic unit.
```
Comment (lukechurch) [inline webapp/lib/model.g.dart]: They're not atomic - that would only be the case if we were using transactions - could you adjust the comments here  to 'batch'?

File: webapp/lib/model.g.dart [hunk 2]
```diff
@@ -417,12 +417,43 @@ void listenForUpdates<T>(
...
+  /// Note that it won't resolve while you're offline.
+  Future<Null> commit();
+
+  /// Updates fields in the document referred to by this [DocumentReference].
+  /// The update will fail if applied to a document that does not exist.
+  void update(firestore.DocumentReference doc, {Map<String, dynamic> data});
+}
+
+/// A batch update for documents in firestore.
+class FsDocUpdate implements DocUpdate {
```
Comment (lukechurch) [inline webapp/lib/model.g.dart]: I think we need to expand 'Fs' - It mean 'filesystem' to me first off. Could we Go with the full verbose thing of FireStore?

File: webapp/lib/platform.dart [hunk 3]
```diff
@@ -138,13 +138,13 @@ Future updateUnread(List<Conversation> conversations, bool newValue) async {
       : "${conversations.length} conversations"
   }");
   if (conversations.isEmpty) return null;
-  var batch = _firestoreInstance.batch();
+  var batch = FsDocUpdate(_firestoreInstance.batch());
   int batchSize = 0;
```
Comment (lukechurch) [inline webapp/lib/platform.dart]: Where does batch size get updated here? It seems to always be zero?

File: webapp/lib/model.g.dart [hunk 4]
```diff
@@ -417,12 +417,43 @@ void listenForUpdates<T>(
...
   final String id;
   final Map<String, dynamic> data;
 
   DocSnapshot(this.id, this.data);
 }
 
+/// A batch update, used to perform multiple writes as a single atomic unit.
+/// None of the writes are committed (or visible locally) until
+/// [DocUpdate.commit()] is called.
+abstract class DocUpdate {
```
Comment (lukechurch) [inline webapp/lib/model.g.dart]: I wonder if this should be called `DocUpdateBatch`? It wasn't obvious to me from the name that it was really doing batching

File: webapp/lib/model.g.dart (hunk 1 above)
Comment (danrubel) [reply webapp/lib/model.g.dart]: @lukechurch  According to the [WriteBatch online docs](https://firebase.google.com/docs/reference/android/com/google/firebase/firestore/WriteBatch), it is atomic.

Please let me know how you would like to proceed and I can update the comment in a future PR.

File: webapp/lib/model.g.dart (hunk 2 above)
Comment (danrubel) [reply webapp/lib/model.g.dart]: Good point. Done.

File: webapp/lib/platform.dart (hunk 3 above)
Comment (danrubel) [reply webapp/lib/platform.dart]: Thanks! Good catch! Fixed.

File: webapp/lib/model.g.dart (hunk 4 above)
Comment (danrubel) [reply webapp/lib/model.g.dart]: Good point. `DocUpdateBatch` is more indicative of its function. Renamed.

**Pull request merged**